File: packages/csharp/Ordex.WhatsApp.Sdk/src/Ordex.WhatsApp.Sdk/Entitlement/Entitlement.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;

namespace Ordex.WhatsApp.Sdk.Entitlement;

public sealed class StaticEntitlementChecker : IEntitlementChecker
{
    private readonly bool _enabled;
    private readonly string? _plan;
    private readonly string? _message;

    public StaticEntitlementChecker(bool enabled, string? plan = "saas", string? message = null)
    {
        _enabled = enabled;
        _plan = plan;
        _message = message;
    }

    public Task<EntitlementResult> CheckAsync(string tenantId, CancellationToken cancellationToken = default) =>
        Task.FromResult(new EntitlementResult(_enabled, tenantId, "whatsapp", _plan ?? "saas", _message));
}

/// <summary>
/// Checagem plugável contra Ordex Pay.
/// 1. POST /addons/whatsapp/entitlement
/// 2. Se 404, fallback POST /licenses/verify e lê addons.whatsapp.
/// </summary>
public sealed class OrdexPayEntitlementChecker : IEntitlementChecker
{
    private readonly HttpClient _http;
    private readonly string _apiKey;
    private readonly bool _treatValidLicenseAsAddon;
    private readonly string? _clientId;

    public OrdexPayEntitlementChecker(
        string apiKey,
        HttpClient http,
        string? clientId = null,
        bool treatValidLicenseAsAddon = false)
    {
        _apiKey = apiKey;
        _http = http;
        _clientId = clientId;
        _treatValidLicenseAsAddon = treatValidLicenseAsAddon;
        if (_http.BaseAddress is null)
        {
            _http.BaseAddress = new Uri(AgendaCobranca.Sdk.AgendaCobrancaOptions.DefaultBaseUrl.TrimEnd('/') + "/");
        }
    }

    public async Task<EntitlementResult> CheckAsync(string tenantId, CancellationToken cancellationToken = default)
    {
        try
        {
            var payload = await PostAsync("addons/whatsapp/entitlement", new
            {
                tenant_id = tenantId,
                client_id = _clientId ?? ""
            }, cancellationToken).ConfigureAwait(false);
            return FromPayload(tenantId, payload, "addons/whatsapp/entitlement");
        }
        catch (WhatsAppNotFoundException)
        {
        }

        var license = await PostAsync("licenses/verify", new
        {
            client_id = _clientId ?? "",
            tenant_id = tenantId
        }, cancellationToken).ConfigureAwait(false);
        return FromLicense(tenantId, license);
    }

    internal static async Task<EntitlementResult> AssertEnabledAsync(
        IEntitlementChecker? checker,
        string tenantId,
        bool skip,
        CancellationToken cancellationToken = default)
    {
        if (skip)
        {
            return new EntitlementResult(true, tenantId, "whatsapp", "demo", "skipEntitlementCheck");
        }

        if (checker is null)
        {
            throw new WhatsAppEntitlementException(
                "EntitlementChecker ausente: injete StaticEntitlementChecker (demo) ou OrdexPayEntitlementChecker",
                tenantId);
        }

        var result = await checker.CheckAsync(tenantId, cancellationToken).ConfigureAwait(false);
        if (!result.Enabled)
        {
            throw new WhatsAppEntitlementException(
                result.Message ??
                "Add-on WhatsApp do plano SaaS Ordex Pay nao esta habilitado para este tenant",
                tenantId);
        }

        return result;
    }

    private async Task<JsonElement> PostAsync(string path, object body, CancellationToken cancellationToken)
    {
        using var message = new HttpRequestMessage(HttpMethod.Post, path)
        {
            Content = new StringContent(JsonSerializer.Serialize(body, JsonOptions.Default), Encoding.UTF8, "application/json")
        };
        message.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
        message.Headers.TryAddWithoutValidation("chave_api", _apiKey);
        message.Headers.TryAddWithoutValidation("X-Api-Key", _apiKey);

        using var response = await _http.SendAsync(message, cancellationToken).ConfigureAwait(false);
        var raw = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
        if (response.IsSuccessStatusCode)
        {
            return Parse(raw);
        }

        var rec = Parse(raw);
        var msg = rec.TryGetProperty("message", out var m) ? m.GetString()
            : rec.TryGetProperty("error", out var e) ? e.GetString()
            : $"Erro HTTP {(int)response.StatusCode}";
        throw ErrorFromStatus.Create((int)response.StatusCode, msg ?? $"Erro HTTP {(int)response.StatusCode}", raw);
    }

    private static EntitlementResult FromPayload(string tenantId, JsonElement payload, string source)
    {
        var dados = Unwrap(payload);
        var enabled = IsTruthy(dados, "enabled") || IsTruthy(dados, "valid") ||
                      IsTruthy(payload, "enabled") || IsTruthy(payload, "valid") ||
                      AddonFlag(dados) || AddonFlag(payload);
        var message = GetString(dados, "message") ?? GetString(payload, "message") ?? source;
        var plan = GetString(dados, "plan") ?? GetString(payload, "plan") ?? "saas";
        return new EntitlementResult(enabled, tenantId, "whatsapp", plan, message, payload.Clone());
    }

    private EntitlementResult FromLicense(string tenantId, JsonElement payload)
    {
        var dados = Unwrap(payload);
        var licenseValid = IsTruthy(dados, "valid") || IsTruthy(payload, "valid") || IsTruthy(payload, "success");
        var addon = AddonFlag(dados) || AddonFlag(payload);
        var enabled = addon || (_treatValidLicenseAsAddon && licenseValid);
        var message = enabled
            ? "add-on whatsapp autorizado via licenses/verify"
            : "licenses/verify sem add-on whatsapp habilitado";
        return new EntitlementResult(enabled, tenantId, "whatsapp", "saas", message, payload.Clone());
    }

    private static JsonElement Unwrap(JsonElement payload)
    {
        if (payload.ValueKind == JsonValueKind.Object && payload.TryGetProperty("data", out var data) &&
            data.ValueKind == JsonValueKind.Object)
        {
            return data;
        }

        return payload;
    }

    private static bool AddonFlag(JsonElement rec)
    {
        if (rec.ValueKind != JsonValueKind.Object) return false;
        if (IsTruthy(rec, "whatsapp_addon") || IsTruthy(rec, "addon_whatsapp") || IsTruthy(rec, "whatsapp"))
        {
            return true;
        }

        if (!rec.TryGetProperty("addons", out var addons) || addons.ValueKind != JsonValueKind.Object)
        {
            return false;
        }

        if (!addons.TryGetProperty("whatsapp", out var wa)) return false;
        if (IsTruthyValue(wa)) return true;
        if (wa.ValueKind == JsonValueKind.Object)
        {
            return IsTruthy(wa, "enabled") || IsTruthy(wa, "valid");
        }

        return false;
    }

    private static bool IsTruthy(JsonElement rec, string name) =>
        rec.ValueKind == JsonValueKind.Object && rec.TryGetProperty(name, out var el) && IsTruthyValue(el);

    private static bool IsTruthyValue(JsonElement el) =>
        el.ValueKind == JsonValueKind.True ||
        (el.ValueKind == JsonValueKind.Number && el.TryGetInt32(out var n) && n == 1) ||
        (el.ValueKind == JsonValueKind.String && (el.GetString() is "true" or "1"));

    private static string? GetString(JsonElement rec, string name) =>
        rec.ValueKind == JsonValueKind.Object && rec.TryGetProperty(name, out var el) && el.ValueKind == JsonValueKind.String
            ? el.GetString()
            : null;

    private static JsonElement Parse(string raw)
    {
        if (string.IsNullOrWhiteSpace(raw)) return JsonDocument.Parse("{}").RootElement.Clone();
        try
        {
            return JsonDocument.Parse(raw).RootElement.Clone();
        }
        catch (JsonException)
        {
            return JsonDocument.Parse($"{{\"raw\":{JsonSerializer.Serialize(raw)}}}").RootElement.Clone();
        }
    }
}

File: packages/ruby/whatsapp/lib/ordex_whatsapp/entitlement.rb
# frozen_string_literal: true

require "faraday"
require "json"

module OrdexWhatsApp
  class StaticEntitlementChecker
    def initialize(enabled, plan: "saas", message: nil)
      @enabled = enabled
      @plan = plan
      @message = message
    end

    def check(tenant_id)
      { enabled: @enabled, tenant_id: tenant_id, add_on: "whatsapp", plan: @plan, message: @message }
    end
  end

  class OrdexPayEntitlementChecker
    def initialize(api_key, base_url: nil, connection: nil, treat_valid_license_as_addon: false)
      @api_key = api_key
      @base_url = (base_url.nil? || base_url.to_s.strip.empty? ? DEFAULT_ORDEX_BASE_URL : base_url).to_s.sub(%r{/$}, "")
      @connection = connection || Faraday.new(url: "#{@base_url}/") do |f|
        f.request :json
        f.adapter Faraday.default_adapter
      end
      @treat_valid_license_as_addon = treat_valid_license_as_addon
    end

    def check(tenant_id)
      payload = post("addons/whatsapp/entitlement", tenant_id: tenant_id, client_id: "")
      from_payload(tenant_id, payload, "addons/whatsapp/entitlement")
    rescue NotFoundError
      license = post("licenses/verify", client_id: "", tenant_id: tenant_id)
      from_license(tenant_id, license)
    end

    def self.assert!(checker, tenant_id, skip:)
      return { enabled: true, tenant_id: tenant_id, add_on: "whatsapp", plan: "demo", message: "skipEntitlementCheck" } if skip
      if checker.nil?
        raise EntitlementError.new(
          "EntitlementChecker ausente: injete StaticEntitlementChecker (demo) ou OrdexPayEntitlementChecker",
          tenant_id: tenant_id
        )
      end
      result = symbolize(checker.check(tenant_id))
      unless result[:enabled]
        raise EntitlementError.new(
          result[:message] || "Add-on WhatsApp do plano SaaS Ordex Pay nao esta habilitado para este tenant",
          tenant_id: tenant_id
        )
      end
      result
    end

    private

    def post(path, body)
      resposta = @connection.post(path) do |req|
        req.headers["Accept"] = "application/json"
        req.headers["Content-Type"] = "application/json"
        req.headers["chave_api"] = @api_key
        req.headers["X-Api-Key"] = @api_key
        req.body = JSON.dump(body)
      end
      payload = decode(resposta.body)
      return payload if resposta.success?

      rec = payload.is_a?(Hash) ? payload : {}
      mensagem = rec["message"] || rec["error"] || "Erro HTTP #{resposta.status}"
      raise OrdexWhatsApp.error_from_status(resposta.status, mensagem, body: payload)
    end

    def from_payload(tenant_id, payload, source)
      rec = as_hash(payload)
      dados = unwrap(payload)
      enabled = truthy?(dados["enabled"]) || truthy?(dados["valid"]) || truthy?(rec["enabled"]) ||
                truthy?(rec["valid"]) || addon_flag?(dados) || addon_flag?(rec)
      {
        enabled: enabled,
        tenant_id: tenant_id,
        add_on: "whatsapp",
        plan: dados["plan"] || rec["plan"] || "saas",
        message: dados["message"] || rec["message"] || source,
        raw: payload
      }
    end

    def from_license(tenant_id, payload)
      rec = as_hash(payload)
      dados = unwrap(payload)
      license_valid = truthy?(dados["valid"]) || truthy?(rec["valid"]) || truthy?(rec["success"])
      addon = addon_flag?(dados) || addon_flag?(rec)
      enabled = addon || (@treat_valid_license_as_addon && license_valid)
      {
        enabled: enabled,
        tenant_id: tenant_id,
        add_on: "whatsapp",
        plan: "saas",
        message: enabled ? "add-on whatsapp autorizado via licenses/verify" : "licenses/verify sem add-on whatsapp habilitado",
        raw: payload
      }
    end

    def unwrap(payload)
      rec = as_hash(payload)
      data = rec["data"]
      data.is_a?(Hash) ? data : rec
    end

    def as_hash(payload)
      payload.is_a?(Hash) ? payload : {}
    end

    def addon_flag?(rec)
      return false unless rec.is_a?(Hash)
      return true if truthy?(rec["whatsapp_addon"]) || truthy?(rec["addon_whatsapp"]) || truthy?(rec["whatsapp"])

      addons = rec["addons"]
      return false unless addons.is_a?(Hash)

      wa = addons["whatsapp"]
      return true if truthy?(wa)
      return truthy?(wa["enabled"]) || truthy?(wa["valid"]) if wa.is_a?(Hash)

      false
    end

    def truthy?(value)
      value == true || value == "true" || value == 1 || value == "1"
    end

    def decode(corpo)
      return {} if corpo.nil? || corpo.to_s.strip.empty?

      JSON.parse(corpo)
    rescue JSON::ParserError
      { "raw" => corpo }
    end

    def self.symbolize(hash)
      return hash unless hash.is_a?(Hash)

      hash.each_with_object({}) { |(k, v), acc| acc[k.to_sym] = v }
    end
  end
end
